### ISSUE TITLE
Using custom runtime breaks dynamic routes

### ISSUE THREAD

**Issue by capJavert** [vercel/next.js#56869]
### Link to the code that reproduces this issue

https://github.com/capJavert/nextjs-functions-break-dynamic-routes

### To Reproduce

1. Add serverless function with custom runtime (eg. [Go](https://vercel.com/docs/functions/serverless-functions/runtimes/go)) to your Next.js project
2. Add dynamic API routes to your Next.js app inside app directory (or old school dynamic routes through `/api/pages`
3. Deploy to Vercel

You will observe:
- custom runtime handler works [/api/hello](https://nextjs-functions-break-dynamic-routes-bp54lnl4k-capjavert.vercel.app/api/hello) ✅ 
- dynamic route [/api/users/:id](https://nextjs-functions-break-dynamic-routes-bp54lnl4k-capjavert.vercel.app/api/users/1) (created in app directory) does not work and returns 404 🔴 
- dynamic route [/api/posts/:id](https://nextjs-functions-break-dynamic-routes-bp54lnl4k-capjavert.vercel.app/api/posts/1) (created in pages directory) does not work and returns 404 🔴 

I also noticed api routes without dynamic segments work fine [/api/boo](https://nextjs-functions-break-dynamic-routes-bp54lnl4k-capjavert.vercel.app/api/boo).

Deployment summary also shows all the functions correctly:
<img width="1177" alt="image" src="https://github.com/vercel/next.js/assets/9803078/675f07b8-37ce-4e4e-8a4a-9b5e465ad2f5">

### Current vs. Expected behavior

When custom runtime is used inside top level `/api` directory Next.js dynamic routes break and show 404. 

Dynamic routes should work as expected.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.0.0: Fri Sep 15 14:41:43 PDT 2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T6000
Binaries:
  Node: 18.16.0
  npm: 9.5.1
  Yarn: 1.22.18
  pnpm: 8.6.3
Relevant Packages:
  next: 13.5.5-canary.16
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Middleware / Edge (API routes, runtime)

**Comment by barisceliker1** [comment]
I also have the same problem. Did you find anything ?

**Comment by capJavert** [comment]
Nope, just broken.

**Comment by yoeven** [comment]
Same issue here, broke an entire prod project and spent hours trying to fix it. Works well locally but not when deployed,  one hack that fixed it was using a subdomain with rewrites to the api folder from the next.config.js. However, domains without a rewrite to the API folder breaks

**Comment by capJavert** [comment]
I can confirm this is still issue in 14.1. But unfortunately like many crucial bugs like this no response still.

**Comment by capJavert** [comment]
This is still happening in latest release. Any reply/ack?

**Comment by capJavert** [comment]
Vercel can we get any update for this issue, it has been months now and it is still not fixed. I tested on the latest version and it is still happening.

**Comment by capJavert** [comment]
I am a creator of this issue, why are my comments where I ask for updates marked as off topic, don't do it maintainer bot.

**Comment by codeXLinkX** [comment]
Facing the same issue.

**Comment by capJavert** [comment]
@leerob can we get some feedback here and also your maintainer bot is running rogue and muting people comments. If there is another door to knock to instead of tagging you, just say, thank you!

**Comment by codeXLinkX** [comment]
Facing the same issue.

I am able to reproduce the same issue with a python runtime file in api/index.py.

When I check the Vercel deployment summary, I can see that it was able to bundle the dynamic route /api/test/[slug] under the functions, but when it tries to call that route, I get 404.

**Comment by codeXLinkX** [comment]
Ok I just figured it out I think.

In my case it is a Python FastAPI setup with existing NextJS routes, following the Next.js FastAPI Starter example
https://github.com/digitros/nextjs-fastapi.

The above example didn't contain any dynamic routes, that's why the rewrites there is incomplete.

The line `{ source: "/api/:path*", destination: "/api/:path*" }` has to be added to the rewrites, to avoid breaking the dynamic routes.

`const nextConfig = {
  reactStrictMode: false,
  transpilePackages: ["react-pdf"],
  rewrites: async () => {
    return [
      {
        source: "/api/py/:path*",
        destination:
          process.env.NODE_ENV === "development"
            ? "http://127.0.0.1:8000/api/py/:path*"
            : "/api/",
      },
      { source: "/api/:path*", destination: "/api/:path*" },
    ];
  },
};`

@capJavert you probably need to do something similar for Go.

**Comment by capJavert** [comment]
Yes, but that is not a bug fix, it is just a temp avoidance of the issue.

**Comment by yousefamar** [comment]
Wow, hours wasted! Thanks for sharing the workaround @capJavert 

As another suggestion, since this only happens when there's a root-level `/api` folder: maybe Vercel could allow configuring the name of that folder (for Python runtime). Two birds one stone as I never liked having to call it `/api` in the first place.

**Comment by marcoshernanz** [comment]
This was fixed in [vercel/vercel#15297](https://github.com/vercel/vercel/pull/15297). I redeployed the original reproduction and verified that the Go function, static API route, App Router dynamic route, and Pages Router dynamic route all return 200 on the reported Next.js version, current stable, and current canary. Closing as fixed. Thanks for the report!
